include lock file and use hex packages where available

tsloughter opened the pull request:
I'd like to get this published to hex and this is the first step in that direction. It will require publishing `parse_xfrm_utils` and `dns_erlang` first but I thought I'd send this PR with some initial changes needed, and some changes to be more idiomatic rebar3 usage.

tsloughter commented:
@aeden I'm not sure which changes are unrelated?

aeden commented:
These bits jump out at me:

https://github.com/dnsimple/erldns/pull/90/files#diff-d0d06d8f5a21085d7390330b06d15f8b 
https://github.com/dnsimple/erldns/pull/90/files#diff-974eee8df153ea55ef883816b18e6d5e
removal of test.sh and run.sh

Perhaps I've misunderstood why the changes are necessary though to publish to hex.

Additionally there are changes that may no longer be relevant because master is now on OTP 21.

tsloughter commented:
Oh, I see I didn't update my original post about this. All deps have to also be hex dependencies, that is why I removed `parse_xfrm_utils` which I thought was simpler than getting it published.

The script files could be kept and simply changed I suppose. I figured since `skip-deps` is no longer needed there wasn't a need for `test.sh` as the command becomes as simple as `rebar3 eunit`. But I can add it back.

For `run.sh` it would be a script that simply runs `rebar3 shell`.

aeden commented:
If you'd like to resolve the conflicts, I can test it out in our environment and make sure that it does not cause any issues for us, and merge it if so.

tsloughter commented:
Ok, updated.

Note that I removed `get-deps` from the makefile. It is not needed and only meant for cases similar to vendoring, like cases that a build requires first fetching dependencies and then compile being run without network access.

tsloughter commented:
Side note, it is a bug in rebar3 if `make fresh` is ever actually needed. rebar3 is meant to be deterministic with dependencies even during upgrades (unlike rebar2) and is will get the correct deps if the lock file changes (like when a developer changes branches and on the new branch a different set of deps is currently used and in the lock file).

aeden commented on rebar.config:
```diff
@@ -1,33 +1,29 @@
+{profiles, [{test, [{deps, [proper]}]}]}.
+
+{shell, [{apps, [erldns]},
+         {config, "erldns.config.example"}]}.
```
Any objections to changing this to "erldns.config" rather than using the example config?

tsloughter replied:
Should I rename the file? This is based on what you have in `run.sh` which uses the example config.

aeden replied:
No, just use `erldns.config` here, as it was in run.sh: https://github.com/dnsimple/erldns/pull/90/files#diff-0a369498a5a8db3ac8fa606b544c9810L36

tsloughter replied:
Ooh, so the idea is that the user copy it to `erldns.config` first?

aeden replied:
Exactly.

aeden replied:
If you want I can merge and then make this change post merge.

tsloughter replied:
I just updated it.

aeden replied:
👍

aeden commented on rebar.config:
```diff
@@ -1,33 +1,29 @@
+        recon,
+        folsom,
+        jsx,
         {dns, ".*", {git, "git://github.com/dnsimple/dns_erlang.git", {branch, "master"}}},
```
Does it make sense to try to get dns_erlang into hex as well?

tsloughter replied:
Yup, I have a PR related to that open as well.

tsloughter replied:
This may be of interest, all this started because we are using erldns in the example project (https://github.com/adoptingerlang/service_discovery) for https://adoptingerlang.org/

aeden commented:
Thank you @tsloughter for making this happen, I appreciate it.
